compiler/rustc_typeck/src/check/wfcheck.rs

```diff
@@ -692,19 +686,12 @@ impl<'tcx> GATSubstCollector<'tcx> {
 impl<'tcx> TypeVisitor<'tcx> for GATSubstCollector<'tcx> {
     type BreakTy = !;
 
-    fn visit_binder<T: TypeFoldable<'tcx>>(
-        &mut self,
-        t: &ty::Binder<'tcx, T>,
-    ) -> ControlFlow<Self::BreakTy> {
-        self.tcx.liberate_late_bound_regions(self.gat, t.clone()).visit_with(self)
-    }
-
     fn visit_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<Self::BreakTy> {
         match t.kind() {
             ty::Projection(p) if p.item_def_id == self.gat => {
                 for (idx, subst) in p.substs.iter().enumerate() {
                     match subst.unpack() {
-                        GenericArgKind::Lifetime(lt) => {
+                        GenericArgKind::Lifetime(lt) if !lt.is_late_bound() => {
```

Comment:
I'm not sure this is completely correct. We can't skip late-bound regions on the *function*.  (But, we liberate those right?)

Reply:
So we can't just skip late-bound regions then (I was just going by [this comment](https://rust-lang.zulipchat.com/#narrow/stream/144729-t-types/topic/Aligning.20rustc.20and.20chalk.20folding.2Fvisiting.20traits/near/286518529) in our discussion yesterday)?

In that case, is [my previous effort](https://rust-lang.zulipchat.com/#narrow/stream/144729-t-types/topic/Aligning.20rustc.20and.20chalk.20folding.2Fvisiting.20traits/near/286517000) correct?  When visiting a late-bound region it just delegates into visiting a liberated version thereof.

Or are there some late-bound regions that *should* be skipped entirely?

Reply:
Let me look real quick. I think its fine, because I think we liberate the function late-bound regions. We need to skip *inner* late-bound regions though.

Reply:
We do: https://github.com/rust-lang/rust/blob/cdcc53b7dc002ea4a7a28105010c5a1126ee31b7/compiler/rustc_typeck/src/check/wfcheck.rs#L317

This is fine.